### DISCUSSION TITLE
Questions about converting structures to contacts

### DISCUSSION THREAD

**Discussion by LittletreeZou** [facebookresearch/esm#434] [Q&A] [answered]
Thank you for this wonderful work. But I have difficulty understanding the code that converts structures to contacts in https://github.com/facebookresearch/esm/blob/main/examples/contact_prediction.ipynb.  

According to the definition, a pair of residues is defined as a contact if the distance between their Cβ atoms is less than or equal to 8Å. My questions are:
1. Why not directly compute the Cβ atom distance using their coordinates but compute them through the coordinates of N, CA, C?
2. In line `Cbeta = extend(C, N, CA, 1.522, 1.927, -2.143)`, why the last three are fixed constants? Are they the same for all residues?

```
# Converting structures to contacts

def extend(a, b, c, L, A, D):
    """
    input:  3 coords (a,b,c), (L)ength, (A)ngle, and (D)ihedral
    output: 4th coord
    """

    def normalize(x):
        return x / np.linalg.norm(x, ord=2, axis=-1, keepdims=True)

    bc = normalize(b - c)
    n = normalize(np.cross(b - a, bc))
    m = [bc, np.cross(n, bc), n]
    d = [L * np.cos(A), L * np.sin(A) * np.cos(D), -L * np.sin(A) * np.sin(D)]
    return c + sum([m * d for m, d in zip(m, d)])


def contacts_from_pdb(
    structure: bs.AtomArray,
    distance_threshold: float = 8.0,
    chain: Optional[str] = None,
) -> np.ndarray:
    mask = ~structure.hetero
    if chain is not None:
        mask &= structure.chain_id == chain

    N = structure.coord[mask & (structure.atom_name == "N")]
    CA = structure.coord[mask & (structure.atom_name == "CA")]
    C = structure.coord[mask & (structure.atom_name == "C")]

    Cbeta = extend(C, N, CA, 1.522, 1.927, -2.143)
    dist = squareform(pdist(Cbeta))
    
    contacts = dist < distance_threshold
    contacts = contacts.astype(np.int64)
    contacts[np.isnan(dist)] = -1
    return contacts
```

**Comment by tomsercu** [comment, marked as answer]
Glycine has no carbon beta, so this method allows us a consistent approach that applies to all amino acids equally.
The backbone atoms for every single residue form a "rigid body' and those  (L)ength, (A)ngle, and (D)ihedral are constant across all amino acids.

**Reply by LittletreeZou** [reply]
Thanks for your prompt reply!
